android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/lifecyclemanager/AndroidRunner.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.lifecyclemanager;
+
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.util.Exec;
+import com.aws.greengrass.util.Pair;
+import com.aws.greengrass.util.ProxyUtils;
+import com.aws.greengrass.util.platforms.Platform;
+import com.aws.greengrass.util.platforms.android.AndroidCallable;
+import com.aws.greengrass.util.platforms.android.AndroidCallableExec;
+import com.aws.greengrass.util.platforms.android.AndroidShellExec;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+import static com.aws.greengrass.android.managers.AndroidBaseComponentManager.RUN_SERVICE_CMD;
+import static com.aws.greengrass.android.managers.AndroidBaseComponentManager.SHUTDOWN_SERVICE_CMD;
+import static com.aws.greengrass.android.managers.AndroidBaseComponentManager.STARTUP_SERVICE_CMD;
+import static com.aws.greengrass.android.managers.AndroidBasePackageManager.APK_INSTALL_CMD;
+import static com.aws.greengrass.ipc.AuthenticationHandler.SERVICE_UNIQUE_ID_KEY;
+import static com.aws.greengrass.util.Utils.isEmpty;
+
+public class AndroidRunner extends ShellRunner.Default {
+
+    @Override
+    public synchronized Exec setup(String note, String command, GreengrassService onBehalfOf)
+            throws IOException {
+        if (!isEmpty(command) && onBehalfOf != null) {
+            String serviceName = onBehalfOf.getServiceName();
+            Path cwd = nucleusPaths.workPath(serviceName);
+            Logger logger = getLoggerToUse(onBehalfOf);
+            Exec exec = createSpecializedExec(command, serviceName, logger)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *createSpecializedExec* method in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when cd throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).